internal/cfgmgr/cfgmgr.go

```diff
@@ -0,0 +1,150 @@
+package cfgmgr
+
+import (
+	"fmt"
+	"strings"
+	"unicode/utf16"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/pkg/errors"
+)
+
+//go:generate go run ../../mksyscall_windows.go -output zsyscall_windows.go cfgmgr.go
+
+//sys cmGetDeviceIDListSize(pulLen *uint32, pszFilter *byte, uFlags uint32) (hr error) = cfgmgr32.CM_Get_Device_ID_List_SizeA
+//sys cmGetDeviceIDList(pszFilter *byte, buffer *byte, bufferLen uint32, uFlags uint32) (hr error)= cfgmgr32.CM_Get_Device_ID_ListA
+//sys cmLocateDevNode(pdnDevInst *uint32, pDeviceID string, uFlags uint32) (hr error) = cfgmgr32.CM_Locate_DevNodeW
+//sys cmGetDevNodeProperty(dnDevInst uint32, propertyKey *devPropKey, propertyType *uint32, propertyBuffer *uint16, propertyBufferSize *uint32, uFlags uint32) (hr error) = cfgmgr32.CM_Get_DevNode_PropertyW
+
+type devPropKey struct {
+	fmtid guid.GUID
+	pid   uint32
+}
+
+const (
+	_CM_GETIDLIST_FILTER_BUSRELATIONS = uint32(0x00000020)
+
+	_CM_LOCATE_DEVNODE_NORMAL = uint32(0x00000000)
+
+	_DEVPROP_TYPE_STRING        = uint32(0x00000012)
+	_DEVPROP_TYPEMOD_LIST       = uint32(0x00002000)
+	_DEVPROP_TYPE_STRING_LIST   = uint32(_DEVPROP_TYPE_STRING | _DEVPROP_TYPEMOD_LIST)
+	_DEVPKEY_LOCATIONPATHS_GUID = "a45c254e-df1c-4efd-8020-67d146a850e0"
+)
+
+// getDevPKeyDeviceLocationPaths creates a DEVPROPKEY struct for the
+// DEVPKEY_Device_LocationPaths property as defined in devpkey.h
+func getDevPKeyDeviceLocationPaths() (*devPropKey, error) {
+	guid, err := guid.FromString(_DEVPKEY_LOCATIONPATHS_GUID)
+	if err != nil {
+		return nil, err
+	}
+	return &devPropKey{
+		fmtid: guid,
+		// pid value is defined in devpkey.h
+		pid: 37,
+	}, nil
+}
+
+func GetDeviceLocationPathsFromIDs(ids []string) ([]string, error) {
+	result := []string{}
+	devPKeyDeviceLocationPaths, err := getDevPKeyDeviceLocationPaths()
+	if err != nil {
+		return nil, err
+	}
+	for _, id := range ids {
+		var devNodeInst uint32
+		err = cmLocateDevNode(&devNodeInst, id, _CM_LOCATE_DEVNODE_NORMAL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to locate device node for %s", id)
+		}
+		propertyType := uint32(0)
+		propertyBufferSize := uint32(0)
+
+		// get the size of the property buffer by querying with a nil buffer and zeroed propertyBufferSize
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, nil, &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get property buffer size of devnode query for %s with", id)
+		}
+
+		// get the property with the resulting propertyBufferSize
+		propertyBuffer := make([]uint16, propertyBufferSize/2)
```

Comment:
propertyBufferSize returns the size of the resulting buffer in bytes, but since we're using the wide version of the api call, each character in the resulting buffer would actually be 2 bytes wide. So the propertyBufferSize should be even.